[anchors] HTLC second level aggregation in the sweeper

halseth opened the pull request:
This PR finishes the changes needed to make the sweeper bundle HTLC second level transaction in a sweep tx, and resign the inputs.

To achieve this we extend the resolvers to include `SignDetails` containing signatures and information needed for when the inputs are signed.

Depends on #4750 
Depends on #4838

Roasbeef commented on contractcourt/contract_resolvers.go:
```diff
@@ -20,6 +20,10 @@ const (
+
+	// secondLevelConfTarget is the confirmation target we'll use when
+	// adding fees to our second-level HTLC transactions.
+	secondLevelConfTarget = 6
```
Good enough for now, but eventually this should start to be a function of the current height and the deadline to expiry (if the HTLC was contested).

Roasbeef commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
 
-	log.Infof("%T(%x): broadcasting second-layer transition tx: %v",
-		h, h.htlc.RHash[:], spew.Sdump(h.htlcResolution.SignedSuccessTx))
+	// If we have non-nil SignDetails, this means that have a 2nd level
```
With this new statement, can move the two cases into distinct methods to allow them to be unit testable in isolation, and also cut down on some of the vertical space here.

halseth replied:
Done!

Roasbeef commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
-		return nil, err
-	}
+		// Wait for the second level transaction to confirm.
+		// TODO(halseth): could use sweeper result?
```
Result of the spend notification?

Roasbeef commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
+		// have re-arranged outputs.
+		txOut := h.htlcResolution.SignedSuccessTx.TxOut[0]
+		index := -1
+		for i, o := range spend.SpendingTx.TxOut {
```
IIRC, there's a method for this in the `input` pacakge.

halseth replied:
Nice, done!

Roasbeef commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
 		}
 
-		h.outputIncubating = true
+		// TODO(halseth): need to checkpoint?
```
Yes, as otherwise, it'll try to re offer the input to the sweeper, which will error out as the output has already been spent (after the max sweep attempts). Alternatively, you can move things around a bit into two functions, and enter them based on spend/conf notification.

halseth replied:
Now checkpointing is added :)

Roasbeef commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
+		// Update the claim outpoint to the one created by the sweeper.
+		claimOutpoint = *op
+
+	// Otherwise we'll publihs the second-level transaction directly and
```
publihs -> publish

Roasbeef commented on contractcourt/htlc_timeout_resolver.go:
```diff
@@ -311,6 +370,7 @@ func (h *htlcTimeoutResolver) Resolve() (ContractResolver, error) {
 	// spend of the output, and make our next move off of that. Depending
 	// on if this is our commitment, or the remote party's commitment,
 	// we'll be watching a different outpoint and script.
+	// TODO(halseth): these details might not be correct after re-signing?
```
Yeah in this case, the logic needs a large overhaul: we need to wait for the spend, of the output, then only then can we use that to register for confirmations below. Alternatively, capture the second value from `SweepInput` above and use that to listen to know once the output has been spent and confirmed on chain. 

So you can possibly abstract away the "output has been spent" logic here, to present the caller at this level with a single channel, that's either the sweeper ntfn channel or this raw spend notification. The logic _after_ the spend comes through should be identical (need to know if it was a timeout or a success spend).

halseth replied:
I think it should be correct now. The `chainDetailsToWatch` will stay the same, as the outpoint spent by the second level stays the same even though the sweeper re-signs it. Now we use that spend to determine the second-level output later.

Roasbeef commented on contractcourt/htlc_timeout_resolver.go:
```diff
@@ -372,13 +432,83 @@ func (h *htlcTimeoutResolver) Resolve() (ContractResolver, error) {
+			return nil, err
+		}
+
+		// TODO(halseth): need to checkpoint?
```
Either you need to checkpoint, then gate of that new state, or make sure all the logic can be replayed above regardless of which transactions has been broadcast/presented/confirmed.

halseth replied:
Checkpointed.

cfromknecht commented on sweep/sweeper.go:
```diff
@@ -833,6 +834,103 @@ func (s *UtxoSweeper) clusterBySweepFeeRate(inputs pendingInputs) []inputCluster
+
+	// Go through each cluster in as, and merge with the next one from bs
+	// if it has at least the fee rate needed.
+	for i := range as {
```
`for a, i := range as`?

cfromknecht commented on sweep/sweeper.go:
```diff
@@ -833,6 +834,103 @@ func (s *UtxoSweeper) clusterBySweepFeeRate(inputs pendingInputs) []inputCluster
+
+			// We can merge.
+			merged := mergeClusters(a, bs[j])
+			finalClusters = append(finalClusters, merged...)
```
what if there are multiple clusters in b that an be merged with _this_ a?

halseth replied:
Then it will be merged with the first one compatible. This is definitely not an optimized version, there are certainly "smarter" ways of finding clusters to merge with, but wanted to keep it simple for now.

cfromknecht commented on sweep/sweeper.go:
```diff
@@ -752,12 +752,86 @@ func (s *UtxoSweeper) bucketForFeeRate(
+			sweepFeeRate += inputFeeRates[op]
+		}
+
+		sweepFeeRate /= chainfee.SatPerKWeight(len(inputs))
```
doesn’t this need to be a weighted average based on the inputs’ actual weight? this method assumes all inputs are of equal weight

halseth replied:
I think this is correct (if we want to average the feerates), since we taking the average fee rate here, not the fee.  This is pre-existing though, I'm just using the same logic as below.

cfromknecht replied:
hmm i'm not so sure, say you have two inputs A (100w, 1000sat/kw) and B (200w, 3000sat/kw). individually, input A would pay 100w*1000sat/kw = 100 sat and B would pay 200w*3000sat/kw = 600sat. combined they pay 700sat total for 300w, or 2.333sat/w.

the unweighted average of the fee rates gives you 2000sat/kw, so total of 300w * 2000sat/kw = 600sat.

however, a weighted average produces (1000sat/kw + 2*3000sat/kw)/3 = 2333sat/kw which gives us the expected fee rate. then 100w*2333sat/kw + 200w*2333sat/kw = 700sat as expected.

> This is pre-existing though, I'm just using the same logic as below.

This may be, but I still think it's incorrect :P

halseth replied:
> which gives us the expected fee rate

Sure, but I don't think it is defined what is the "expected" fee rate here. Here we cluster some inputs with different fee rates togheter, and need to decide which fee rate to use for all of them. Even though one is heavier than the other, I'm not sure it is "more correct" to choose a fee rate that is closer to that one.

I do agree though, that it might be a better strategy. Since many tiny outputs could bring the fee rate down (and since they are tiny they don't cost much anyway) it could be more correct to take the weighted average. You okay with doing this as a follow up (must change it in the pre-existing case as well)?

cfromknecht replied:
okay, maybe it helps to clarify the objective. say you have the weights and fee rate for each input, `(w_i, r_i), i in [1, n]`. my assumption was that this is trying to compute `r_agg` where `sum (w_i * r_i) = r * sum (w_i)`, where `r_agg` represents the _actual_ fee rate of the cluster after aggregating all the desired fees paid by individual inputs.

a different example, it might make sense to take `r_max = max(r_i)`, which then forces all lower fee-rate inputs to pay the same fee-rate as the input with highest priority. unlike `r_agg`, this way creates a `sweepFeeRate` that never decrease the priority of an input, which seems useful if certain inputs are time-sensitive.

i could also see e.g taking the most recent `r_i`, which is maybe viewed as the most up-to-date fee rate from our estimator/user, and forcing all inputs to readjust to that. this value could fluctuate though, so maybe it's not ideal for time-sensitive inputs.

i agree that there are different heuristics we can choose here, and one may not necessarily be more correct. if the goal is simply to pick an arbitrary value between `min(r_i)` and `max(r_i)`, then an unweighted average works. however it still doesn't translate into anything meaningful, which is why imo it seems less correct than some of the other options available.

> You okay with me doing this as a follow up (must change it in the pre-existing case as well)?

Sure! Yeah I'm okay with a follow up :)

halseth replied:
Issue: https://github.com/lightningnetwork/lnd/issues/4812

cfromknecht commented on sweep/tx_input_set.go:
```diff
@@ -83,14 +83,18 @@ type txInputSet struct {
 }
 
+func dustLimit(relayFee chainfee.SatPerKWeight) btcutil.Amount {
+	return txrules.GetDustThreshold(
```
this will be wrong for litecoin..

halseth replied:
oops. Added to the tracking issue: https://github.com/lightningnetwork/lnd/issues/3946

cfromknecht commented on sweep/txgenerator.go:
```diff
@@ -169,14 +169,16 @@ func createSweepTx(inputs []input.Input, outputPkScript []byte,
-		PkScript: outputPkScript,
-		Value:    sweepAmt,
-	})
+	if sweepAmt >= dustLimit {
```
should this be an error?

halseth replied:
If there are no outputs on the tx, it will be caught by the call to `blockchain.CheckTransactionSanity` below.

cfromknecht replied:
maybe we should call `CheckTransactionSanity` inside this method? idk if we can assume every caller will know it needs to do that check manually.

cfromknecht replied:
scratch that i see that it already is :)

cfromknecht commented on sweep/tx_input_set.go:
```diff
@@ -131,28 +189,35 @@ func (t *txInputSet) addToState(inp input.Input, constraints addConstraints) *tx
+	// If the input comes with a required tx out that is below dust, we
+	// won't add it.
+	reqOut := inp.RequiredTxOut()
+	if reqOut != nil && btcutil.Amount(reqOut.Value) < t.dustLimit {
```
Is this possible?

halseth replied:
it depends on whether the relay fee changes, so I would say it's unlikely.

cfromknecht commented on input/size.go:
```diff
@@ -328,6 +328,12 @@ const (
+	//
+	// TODO(halseth): the non-confirmed version currently includes the
+	// overhead.
+	AcceptedHtlcScriptSizeConfirmed = AcceptedHtlcScriptSize // + HtlcConfirmedScriptOverhead
```
intentional comment?

halseth replied:
Yes.

cfromknecht reviewed (commented):
👍 initial pass to familiarize myself with the changes

halseth commented on input/size.go:
```diff
@@ -328,6 +328,12 @@ const (
 
+	// AcceptedHtlcScriptSizeConfirmed 143 bytes
+	//
+	// TODO(halseth): the non-confirmed version currently includes the
```
Yes, that's why the overhead is commented out below. Aiming to fix this in #4775

cfromknecht replied:
iirc we decided having the extra constant wasn't worth it bc we were only overestimating by 3 bytes. what changed?

found the comment: https://github.com/lightningnetwork/lnd/pull/3821/files#r387578921

halseth replied:
Yeah, that's right. I will look into in the other PR whether it is worth keeping it this way, but I felt as we now need a new witnesstype for these inputs, it was better to define a separate weight constants for them.

halseth commented on input/size.go:
```diff
@@ -361,6 +367,12 @@ const (
+	// AcceptedHtlcSuccessWitnessSizeConfirmed 327 bytes
+	//
+	// Input to second level success tx, spending 1 CSV delayed HTLC output.
+	AcceptedHtlcSuccessWitnessSizeConfirmed = 1 + 1 + 1 + 73 + 1 + 73 + 1 + 32 + 1 +
```
Hm, not entirely sure what you mean, but probably a discussion for #4775.

halseth commented:
Pushed new version that does checkpointing, and moved much of the logic into separate methods. PTAL

cfromknecht commented on contractcourt/briefcase.go:
```diff
@@ -980,6 +1045,83 @@ func (b *boltArbitratorLog) checkpointContract(c ContractResolver,
+func encodeSignDetails(w io.Writer, s *input.SignDetails) error {
+	// If we don't have sign details, write false and return.
+	if s == nil {
+		if err := binary.Write(w, endian, false); err != nil {
```
nit: return `binary.Write` directly

halseth replied:
Done!

cfromknecht commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -86,6 +87,31 @@ func (h *htlcSuccessResolver) ResolverKey() []byte {
 
+// waitForSpend waits for the given outpoint to be spent, and returns the
+// details of the spending tx.
+func (h *htlcSuccessResolver) waitForSpend(op *wire.OutPoint,
```
seeing that the previous commit decoupled `waitForHeight` from the resolver, shouldn't we do the same here?

halseth replied:
Done!

cfromknecht commented on contractcourt/htlc_timeout_resolver.go:
```diff
@@ -399,6 +363,59 @@ func (h *htlcTimeoutResolver) Resolve() (ContractResolver, error) {
+
+	// We'll block here until either we exit, or the HTLC output on the
+	// commitment transaction has been spent.
+	spend, err := h.waitForSpend(
```
nit: return `waitForSpend`

halseth replied:
Later commit will add logic below.

cfromknecht reviewed (commented):
latest version looking good! just some minor nits

halseth commented:
Pushed an itest for the HTLC aggregation, and noticed that we cannot rely on the output script when finding the second-level output (since they can be equal for equal HTLCs). Now using the input index instead: https://github.com/lightningnetwork/lnd/compare/b80b30502c9a35fd23128543abb354c531ad0cdc..5e5ab68d384c547e83ac355fe69632e2ecaed08d#diff-b38dbdd572300de530a55756fa7af0afc2935a2a643bdcdef2f12f2253c1d313R356

Here are the full changes since last review (mostly tests and logs in addition to the mentioned bug fix): https://github.com/lightningnetwork/lnd/compare/b80b30502c9a35fd23128543abb354c531ad0cdc..5e5ab68d384c547e83ac355fe69632e2ecaed08d

and addressing the review comments: https://github.com/lightningnetwork/lnd/compare/5e5ab68d384c547e83ac355fe69632e2ecaed08d..f5364880827f01a9e2df7ff71ba1e4852f1898d6

Roasbeef commented on lntest/itest/lnd_multi-hop_htlc_aggregation_test.go:
```diff
@@ -0,0 +1,428 @@
+		carolPayHashes = append(carolPayHashes, payHash[:])
+	}
+
+	// We'll give Alice's invoices a longeger CLTV expiry, to ensure the
```
longeger -> longer

Roasbeef commented on lntest/itest/lnd_multi-hop_htlc_aggregation_test.go:
```diff
@@ -0,0 +1,428 @@
+
+		// If the HTLC has direction towards Alice, Bob will
+		// claim it with the success TX when he learns the
+		// preimage. In this case one extra sat will be on the output, because of the routing fee.
```
Style nit: line fold.

Roasbeef commented on lntest/itest/lnd_multi-hop_htlc_aggregation_test.go:
```diff
@@ -0,0 +1,428 @@
+	// Carol will also sweep her anchor output in a separate tx (since it
+	// will be low fee).
+	if c == commitTypeAnchors {
+		expectedTxes = 4
```
Dem sweet sweet savings

Roasbeef commented on lntest/itest/lnd_multi-hop_htlc_aggregation_test.go:
```diff
@@ -0,0 +1,428 @@
+// resolve them using the second level timeout and success transactions. In
+// case of anchor channels, the second-level spends can also be aggregated and
+// properly feebumped, so we'll check that as well.
+func testMultiHopHtlcAggregation(net *lntest.NetworkHarness, t *harnessTest,
```
Love how comprehensive this test is!

Roasbeef commented:
Mainly only non-blocking comments left, can't wait to get this in! This is such an outstanding PR and really moves `lnd` forward a large step with by finally starting to utilize more of what anchors gives us.

halseth commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -261,6 +166,110 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
+			"incoming+remote htlc confirmed", h,
+			h.htlc.RHash[:])
+
+		// Before we can craft out sweeping transaction, we
```
Done.

Roasbeef commented:
Lingering linter error: 
```
contractcourt/htlc_success_resolver_test.go:87:15: unlambda: replace `func(c ContractResolver,
	r ...*channeldb.ResolverReport) error {
	return testCtx.checkpoint(c, r...)
}` with `testCtx.checkpoint` (gocritic)
		Checkpoint: func(c ContractResolver,
```

Roasbeef reviewed (approved):
LGTM 🧗‍♀️

Should be ready to be rebased and land now! Once in tree, I also plan to run more experiments on testnet to exercise more general functionality and UX from the PoV of a CLI user.

cfromknecht commented on lntest/itest/lnd_test.go:
```diff
@@ -11019,12 +11020,13 @@ func assertActiveHtlcs(nodes []*lntest.HarnessNode, payHashes ...[]byte) error {
 				return fmt.Errorf("node %x didn't have the "+
 					"payHash %v active", node.PubKey[:],
-					payHash)
+					h)
```
can this commit be equally accomplished by replacing `%v` with `%x`? also doesn't incur the 2x memory overhead

halseth replied:
It would, but while debugging I found it useful to just spew the htlcHashed map, so nice to have them hex there IMO.

cfromknecht replied:
seems like a simple change to make in the event one needs to debug, but it's not that important

cfromknecht commented on contractcourt/htlc_timeout_resolver.go:
```diff
@@ -362,14 +441,87 @@ func (h *htlcTimeoutResolver) sweepSecondLevelTransaction(
+		log.Infof("%T(%x): waiting for CSV lock to expire at height %v",
+			h, h.htlc.RHash[:], waitHeight)
+
+		err := waitForHeight(waitHeight, h.Notifier, h.quit)
```
nit: much of the logic below looks duplicated from before, not sure if there is a way to reuse

halseth replied:
The `FindInputIndex` is removed in later commit (will squash with this one), most of the remainder is custom I think.

cfromknecht replied:
yeah was just looking at everything from `watiForHeight` to `SpendInput`, it is very similar but i think it's fine here

cfromknecht commented on contractcourt/htlc_timeout_resolver_test.go:
```diff
@@ -437,3 +438,822 @@ func TestHtlcTimeoutResolver(t *testing.T) {
+		},
+		{
+			// We send a confirmation the sweep tx from published
+			// by the nuresery.
```
nuresery -> nursery

cfromknecht commented on contractcourt/htlc_success_resolver.go:
```diff
@@ -2,7 +2,6 @@ package contractcourt
 
 import (
 	"encoding/binary"
-	"fmt"
```
should this be a fixup (or two) instead?

halseth replied:
Yes, it should! Fixed(up) and squashed!

cfromknecht commented on contractcourt/htlc_timeout_resolver.go:
```diff
@@ -462,28 +462,14 @@ func (h *htlcTimeoutResolver) handleCommitSpend(
 		op := &wire.OutPoint{
 			Hash:  *commitSpend.SpenderTxHash,
-			Index: index,
+			Index: commitSpend.SpenderInputIndex,
```
doesn't the notifier dispatch on scripts rather than inputs?

halseth replied:
Not sure what you mean?

cfromknecht replied:
before we had `FindInputIndex` which looked at the prevout to find the index of the output. The notifier dispatches on scripts, which means we'll end up in a situation where it returns the wrong index if there are duplicates, since (i believe) it'll always stop at the first.

cfromknecht replied:
okay nvm! it looks like the index will be correct since we dispatch on both outpoint and script https://github.com/lightningnetwork/lnd/blob/1a73bc7d7407bcbf43886d9855dd95e747a798b0/chainntnfs/txnotifier.go#L1464L1478

cfromknecht commented:
@halseth okay this looks to be in a good state, needs rebase!

halseth commented:
Rebased.

halseth commented:
We should wait until we have zero-fee secondlevel transactions ready before merging this. Investigating some itest failures still.

halseth commented:
Fixed a few issues that arose when testing the zero-fee HTLC second level channel type:

* Missing `SignDetails` in resolver encoding: https://github.com/lightningnetwork/lnd/pull/4779/commits/b8cc871f10bc9636d84c8539c2ee0e0839fed3b3
* Missing report on resolver restoration from disk: https://github.com/lightningnetwork/lnd/pull/4779/commits/b30a5e0b4720e4bb7aeed78a759230b7eb6c78bd and https://github.com/lightningnetwork/lnd/pull/4779/commits/c3f0af0fc1f6d5dc85e25d5d54582ac411ead27a 
* Itest: sweep output one block earlier for anchor channels: https://github.com/lightningnetwork/lnd/pull/4779/commits/1384174a6e061e8783ea335bbb6fb45b671c27dc

With these fixes the itests pass with the zero-fee HTLC channel tye (separate PR). PTAL.

cfromknecht commented:
@halseth data race https://github.com/lightningnetwork/lnd/pull/4779/checks?check_run_id=1510627461

halseth commented:
Fixed itest deadlock issue (#4851) and itest failures due to how/when HTLCs secondlevels are broadcasted by the sweeper (https://github.com/lightningnetwork/lnd/pull/4779/commits/33efe91e514c5190fae230f61c3f1e721ff2c36a). 

PTAL.

cfromknecht reviewed (approved):
LGTM ⚓️
